java/arcs/core/host/ArcHostContextParticle.kt

```diff
@@ -0,0 +1,255 @@
+ * An implicit [Particle] that lives within the [ArcHost] and used as a utility class to
+ * serialize/deserialize [ArcHostContext] information from [Handle]s. It does not live in an
+ * Arc or participate in normal [Particle] lifecycle.
+ */
+class ArcHostContextParticle : AbstractArcHostParticle() {
+    /**
+     * Given an [ArcId], [hostId], and [ArcHostContext], convert these types to Arc Schema
+     * types, and write them to the appropriate handles. See `ArcHostContext.arcs` for schema
+     * definitions.
+     */
+    suspend fun writeArcHostContext(arcId: String, hostId: String, context: ArcHostContext) {
+        try {
+            val connections = context.particles.flatMap {
+                it.key.handles.map { handle ->
+                    ArcHostParticle_HandleConnections(
+                        arcId,
+                        it.key.particleName,
+                        handle.key, // handleName
+                        handle.value.storageKey.toString(),
+                        handle.value.mode.name,
+                        handle.value.type.tag.name,
+                        handle.value.ttl?.minutes?.toDouble() ?: 0.0
+                    )
+                }
+            }
+            val arcState = ArcHostParticle_ArcHostContext(arcId, hostId, context.arcState.name)
+            val particles = context.particles.map {
+                ArcHostParticle_Particles(
+                    arcId,
+                    it.key.particleName,
+                    it.key.location,
+                    it.value.particleState.name,
+                    it.value.consecutiveFailureCount.toDouble()
+                )
+            }
+
+            // Write ArcHostContext
+            handles.arcHostContext.store(arcState)
+
+            // Write Plan.Particle + ParticleContext
+            handles.particles.clear()
+            particles.forEach { handles.particles.store(it) }
+
+            // Write Plan.HandleConnection
+            handles.handleConnections.clear()
+            connections.forEach { handles.handleConnections.store(it) }
+        } catch (e: Exception) {
+            // TODO: retry?
+            throw IllegalStateException("Unable to serialize $arcId for $hostId", e)
+        }
+    }
+
+    /**
+     * Reads [ArcHostContext] from serialized representation as Arcs Schema types. See
+     * `ArcHostContext.arcs' for Schema definitions. NOTE: This is more complex than it needs
+     * to be because references are not supported yet in schema2kotlin, and so this information
+     * is stored in de-normalized format.
+     */
+    suspend fun readArcHostContext(
+        arcId: String,
+        hostId: String,
+        instantiateParticle: suspend (ParticleIdentifier) -> Particle
+    ): ArcHostContext? {
+        try {
+            val arcStateEntity = handles.arcHostContext.fetch()
+            val particleEntities = handles.particles.fetchAll()
+            val connectionEntities = handles.handleConnections.fetchAll()
```

Comment:
If you switched to proto serialisation instead, this could all just go in the one proto. You wouldn't need separate handles.

Reply:
Discussed in chat, let's do this in a follow on either with kotlinx.serialization, is using collection references when support is landed.